In [2]:
import numpy as np
import pandas as pd
import keras.utils as to_categorical


Iris dataset is the one we are using

In [4]:
dataset=pd.read_csv('/content/Iris.csv')

In [ ]:
dataset

Id  SepalLengthCm  ...  PetalWidthCm         Species
0      1            5.1  ...           0.2     Iris-setosa
1      2            4.9  ...           0.2     Iris-setosa
2      3            4.7  ...           0.2     Iris-setosa
3      4            4.6  ...           0.2     Iris-setosa
4      5            5.0  ...           0.2     Iris-setosa
..   ...            ...  ...           ...             ...
145  146            6.7  ...           2.3  Iris-virginica
146  147            6.3  ...           1.9  Iris-virginica
147  148            6.5  ...           2.0  Iris-virginica
148  149            6.2  ...           2.3  Iris-virginica
149  150            5.9  ...           1.8  Iris-virginica

[150 rows x 6 columns]

In [5]:
x=dataset.iloc[:,1:5].values
y=dataset.iloc[:,5].values

from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()
y1=encoder.fit_transform(y)

y=pd.get_dummies(y1).values

print(y)


[[1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [1 0 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 1 0]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 [0 0 1]
 

In [22]:
from sklearn.model_selection import train_test_split

x_train,x_test,y1_train,y1_test=train_test_split(x,y,test_size=0.2,random_state=0)
print(x_test)

[[5.8 2.8 5.1 2.4]
 [6.  2.2 4.  1. ]
 [5.5 4.2 1.4 0.2]
 [7.3 2.9 6.3 1.8]
 [5.  3.4 1.5 0.2]
 [6.3 3.3 6.  2.5]
 [5.  3.5 1.3 0.3]
 [6.7 3.1 4.7 1.5]
 [6.8 2.8 4.8 1.4]
 [6.1 2.8 4.  1.3]
 [6.1 2.6 5.6 1.4]
 [6.4 3.2 4.5 1.5]
 [6.1 2.8 4.7 1.2]
 [6.5 2.8 4.6 1.5]
 [6.1 2.9 4.7 1.4]
 [4.9 3.1 1.5 0.1]
 [6.  2.9 4.5 1.5]
 [5.5 2.6 4.4 1.2]
 [4.8 3.  1.4 0.3]
 [5.4 3.9 1.3 0.4]
 [5.6 2.8 4.9 2. ]
 [5.6 3.  4.5 1.5]
 [4.8 3.4 1.9 0.2]
 [4.4 2.9 1.4 0.2]
 [6.2 2.8 4.8 1.8]
 [4.6 3.6 1.  0.2]
 [5.1 3.8 1.9 0.4]
 [6.2 2.9 4.3 1.3]
 [5.  2.3 3.3 1. ]
 [5.  3.4 1.6 0.4]]


In [68]:
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD,Adam

model =Sequential()


In [69]:
model.add(Dense(10,input_shape=(4,),activation='sigmoid'))
model.add(Dense(8,activation='sigmoid'))
model.add(Dense(6,activation='sigmoid'))
model.add(Dense(3,activation='softmax'))

model.compile(SGD(lr=0.09),'categorical_crossentropy',metrics=['accuracy'])
model.summary()

Model: "sequential_11"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_96 (Dense)             (None, 10)                50        
_________________________________________________________________
dense_97 (Dense)             (None, 8)                 88        
_________________________________________________________________
dense_98 (Dense)             (None, 6)                 54        
_________________________________________________________________
dense_99 (Dense)             (None, 3)                 21        
Total params: 213
Trainable params: 213
Non-trainable params: 0
_________________________________________________________________


In [70]:
model.fit(x_train,y1_train,epochs=100)
y_pred=model.predict(x_test)

Epoch 1/100
4/4 [==============================] - 0s 3ms/step - loss: 1.1585 - accuracy: 0.3083
Epoch 2/100
4/4 [==============================] - 0s 3ms/step - loss: 1.1289 - accuracy: 0.3083
Epoch 3/100
4/4 [==============================] - 0s 3ms/step - loss: 1.1134 - accuracy: 0.3083
Epoch 4/100
4/4 [==============================] - 0s 4ms/step - loss: 1.1064 - accuracy: 0.3583
Epoch 5/100
4/4 [==============================] - 0s 4ms/step - loss: 1.0987 - accuracy: 0.3250
Epoch 6/100
4/4 [==============================] - 0s 4ms/step - loss: 1.0970 - accuracy: 0.3250
Epoch 7/100
4/4 [==============================] - 0s 3ms/step - loss: 1.0948 - accuracy: 0.3667
Epoch 8/100
4/4 [==============================] - 0s 3ms/step - loss: 1.0947 - accuracy: 0.3667
Epoch 9/100
4/4 [==============================] - 0s 4ms/step - loss: 1.0963 - accuracy: 0.3667
Epoch 10/100
4/4 [==============================] - 0s 3ms/step - loss: 1.0944 - accuracy: 0.3667
Epoch 11/100
4/4 [===========

**ADAM with categorical_crossentropy**


> First I have taken learning rate as 0.01 with 100 epochs at 100th iteration the max accuracy acquired is 98.33 %


> Now with learning rate as 0.04 with 100 epochs the max accuracy acquired is  95.83%


>  Now with learning rate as 0.09 with 100 epochs the max accuracy acquired is  93.33%





**ADAM with mean_squared_error**


> First I have taken learning rate as 0.09 with 100 epochs at 100th iteration the max accuracy acquired is 97.50 %


> Now with learning rate as 0.04 with 100 epochs the max accuracy acquired is 97.50%


>  Now with learning rate as 0.01 with 100 epochs the max accuracy acquired is  98.33%







**SGD with mean_squared_error**


> Here I have taken learning rate as 0.09 with 100 epochs at 100th iteration the max accuracy acquired is 36.67 %


> Now with learning rate as 0.04 with 100 epochs the max accuracy acquired is 36.67%


> Now with learning rate as 0.01 with 100 epochs the max accuracy acquired is 30.83%%







**SGD with categorical_crossentropy**

> Here I have taken learning rate as 0.09 with 100 epochs at 100th iteration the max accuracy acquired is 36.67 %



> Now with learning rate as 0.04 with 100 epochs the max accuracy acquired is 36.67%


> Now with learning rate as 0.01 with 100 epochs the max accuracy acquired is 36.67%




**Conclusion**


> Here we can see that Adam optimizer performs better than SGD


> This is because Adam has faster and efficient convergence rate but this is not applicable to all the values and this performs better at lower learning rate.


> While on the other hand SGD has slower convergence rate so it performs better at higher epoch or learning rate.


> At the end we can that Adam can can used almost everywhere we want faster results still depends on the dataset.










